src/finalisers/system.ts

```diff
@@ -0,0 +1,109 @@
+import Chunk from '../Chunk';
+import { Bundle as MagicStringBundle } from 'magic-string';
+import { OutputOptions } from '../rollup/index';
+
+export default function system (
+	chunk: Chunk,
+	magicString: MagicStringBundle,
+	{ getPath, indentString: t, intro, outro }: {
+		indentString: string;
+		getPath: (name: string) => string;
+		intro: string;
+		outro: string
+	},
+	_options: OutputOptions
+) {
+	const { dependencies, exports } = chunk.getModuleDeclarations();
+
+	const deps = dependencies.map(m => `'${getPath(m.id)}'`);
+
+	const importBindings: string[] = [];
+	let starExcludes: string[];
+	const setters: string[] = [];
+	const varOrConst = chunk.graph.varOrConst;
+
+	dependencies.forEach(dep => {
+		let setter: string[] = [];
+		if (dep.imports) {
+			dep.imports.forEach(specifier => {
+				importBindings.push(specifier.local);
+				if (specifier.imported === '*') {
+					setter.push(`${specifier.local} = module;`);
+				} else {
+					setter.push(`${specifier.local} = module.${specifier.imported};`);
+				}
+			});
+		}
+		if (dep.reexports) {
+			// bulk-reexport form
+			if (dep.reexports.length > 1 || dep.reexports.length === 1 && dep.reexports[0].imported === '*') {
+				setter.push(`${varOrConst} _setter = {};`);
+				// star reexports
+				dep.reexports.forEach(specifier => {
+					if (specifier.imported !== '*')
+						return;
+					// need own exports list for deduping in star export case
+					if (!starExcludes) {
+						starExcludes = exports.map(expt => expt.exported);
+						if (starExcludes.indexOf('default') === -1)
+							starExcludes.push('default');
+						// also include reexport names
+						dependencies.forEach(dep => {
```

Comment:
This is unfortunate naming as `dep` here shadows `dep` in the upper context (which is the kind of thing I hope at some point we can have a linter point out) which hurts readability.

As this block is only executed once and otherwise clutters the loop, I would suggestExtracting it e.g. as a `getStarExcludes(exports, dependencies)` which will also solve the naming conflict and again make the really long function a little more readable 😉

Also it might be worthwhile to consider making `starExcludes` e.g. a `Set` which will make all costly `.indexOf` checks obsolete and can easily be constructed from any array via `new Set(someArray)` (also being a native data structure, Sets are quite optimised for access). Of course in the end, you would need to write something like `Array.from(starExcludes).join(...)` and you would loose guaranteed order which in this case does not seem important (but I may be wrong).

Reply:
Or of course using an object + `Object.keys`